axum/src/lib.rs

```diff
@@ -168,13 +168,48 @@
 //! pool of database connections or clients to other services.
 //!
 //! The two most common ways of doing that are:
+//! - Using the [`State`] extractor.
 //! - Using request extensions
 //! - Using closure captures
 //!
+//! ## Using the [`State`] extractor
+//!
+//! ```rust,no_run
+//! use axum::{
+//!     extract::State,
+//!     routing::get,
+//!     Router,
+//! };
+//! use std::sync::Arc;
+//!
+//! struct AppState {
+//!     // ...
+//! }
+//!
+//! let shared_state = Arc::new(AppState { /* ... */ });
+//!
+//! let app = Router::with_state(shared_state)
+//!     .route("/", get(handler));
+//!
+//! async fn handler(
+//!     State(state): State<Arc<AppState>>,
+//! ) {
+//!     // ...
+//! }
+//! # async {
+//! # axum::Server::bind(&"".parse().unwrap()).serve(app.into_make_service()).await.unwrap();
+//! # };
+//! ```
+//!
+//! You should prefer using [`State`] if possible since it's more type safe. The downside is that
+//! its less dynamic than request extensions.
```

Comment:
I was trying a similar experiment to have an `AnyMap` within `State` to allow using dynamic `Extension` that is not type-safe. But the issue with that is that the impl crashes due to `impl FromRef<T> for State` crashes with those specific implementations of `FromRef`.

I believe this is also more performant since we no longer have to go through a `HashMap` of types (`AnyMap`) and can access the fields directly. Not sure if we want to write it down but I will try to do a benchmark and see what's the difference.

Reply:
We already have a benchmark of extension vs state in the repo.